src/index.c

```diff
@@ -1065,6 +1065,70 @@ static int check_file_directory_collision(git_index *index,
 	return 0;
 }
 
+static int canonicalize_directory_path(
+	git_index *index, git_index_entry *entry)
+{
+	const git_index_entry *match, *best = NULL;
+	char *search, *sep;
+	size_t pos, search_len, best_len;
+
+	if (!index->ignore_case)
+		return 0;
+
+	/* item already exists in the index, simply re-use the existing case */
+	if ((match = git_index_get_bypath(index, entry->path, 0)) != NULL) {
+		memcpy((char *)entry->path, match->path, sizeof(entry->path));
```

Comment:
Funnily enough, the tests which trigger this do have their change within the first 8 bytes :)

However, even making it so this `memcpy` fails to update the whole path, the tests do still succeed. So either we have similar logic elsewhere or the tests are deficient.